parser/prism/Translator.cc

```diff
@@ -2933,6 +2882,76 @@ NodeVec Translator::translateArguments(pm_arguments_node *argsNode, pm_node *blo
     return results;
 }
 
+ast::ExpressionPtr Translator::desugarArray(core::LocOffsets location, absl::Span<pm_node_t *> prismElements,
+                                            ast::Array::ENTRY_store elements) {
+    auto locZeroLen = location.copyWithZeroLength();
+
+    ast::Array::ENTRY_store elems;
+    elems.reserve(elements.size());
+
+    ExpressionPtr lastMerge;
+    ENFORCE(elements.size() == prismElements.size());
+    for (int i = 0; i < elements.size(); i++) {
+        auto *node = prismElements[i];
+        auto &stat = elements[i];
+
+        if (PM_NODE_TYPE_P(node, PM_SPLAT_NODE)) {
+            // Desugar [a, *x, remaining] into a.concat(<splat>(x)).concat(remaining)
+
+            // The Splat was already desugared to Send{Magic.splat(arg)} with the splat's own location.
+            // But for array literals, we want the splat to have the array's location to match
+            // the legacy parser's behavior (important for error messages and hover).
+            auto splatExpr = move(stat);
+
+            // The parser::Send case makes a fake parser::Array with locZeroLen to hide callWithSplat
+            // methods from hover. Using the array's loc means that we will get a zero-length loc for
+            // the Splat in that case, and non-zero if there was a real Array literal.
+            if (auto send = ast::cast_tree<ast::Send>(splatExpr)) {
+                ENFORCE(send->numPosArgs() == 1, "Splat Send should have exactly 1 argument");
+                // Extract the argument from the old Send and create a new one with array's location
+                splatExpr = MK::Splat(location, move(send->getPosArg(0)));
```

Comment:
We can actually move this logic out of this `desugarArray()` helper, and into the `PM_CALL_NODE` code that calls it. Then, the splat locations of regular arrays can remain untampered and correct.

That would fix https://github.com/sorbet/sorbet/issues/9402.